crates/control/src/role_assignment.rs

```diff
@@ -372,289 +285,320 @@ impl RoleAssignment {
+        });
+        let own_network_ball = context.ball_position.map(|seen_ball| {
+            own_ball_to_hulks_network_ball_position(
+                *seen_ball,
+                ground_to_field,
+                context.cycle_time.start_time,
+            )
+        });
+        let ball_position = own_network_ball
+            .or(team_network_ball)
+            .ok_or_eyre("we are striker without a ball, this should never happen")?;
+
+        context
+            .hardware
+            .write_to_network(OutgoingMessage::Spl(HulkMessage::Striker(StrikerMessage {
+                player_number: *context.player_number,
+                pose,
+                ball_position,
+                time_to_reach_kick_position: *context.time_to_reach_kick_position.unwrap(),
+            })))
+            .wrap_err("failed to write StrikerMessage to hardware")
+    }
+
+    fn try_sending_loser_message(
+        &mut self,
+        context: &CycleContext<impl NetworkInterface>,
+        ground_to_field: Isometry2<Ground, Field>,
+    ) -> Result<()> {
+        if !is_enough_message_budget_left(context) {
+            return Ok(());
+        }
+
+        self.last_transmitted_striker_message = Some(context.cycle_time.start_time);
+        self.last_received_spl_striker_message = Some(context.cycle_time.start_time);
+
+        context
+            .hardware
+            .write_to_network(OutgoingMessage::Spl(HulkMessage::Loser(LoserMessage {
+                player_number: *context.player_number,
+                pose: ground_to_field.as_pose(),
+            })))
+            .wrap_err("failed to write LoserMessage to hardware")
+    }
+}
+
+fn is_allowed_to_send_messages(context: &CycleContext<'_, impl NetworkInterface>) -> bool {
+    let is_playing = *context.primary_state == PrimaryState::Playing;
+    let is_penalty_kick =
+        context
+            .filtered_game_controller_state
+            .is_some_and(|game_controller_state| {
+                matches!(
+                    game_controller_state.game_phase,
+                    GamePhase::PenaltyShootout { .. }
+                ) || matches!(game_controller_state.sub_state, Some(SubState::PenaltyKick))
+            });
+
+    is_playing && !is_penalty_kick
+}
+
+fn is_cooldown_elapsed(now: SystemTime, last: Option<SystemTime>, cooldown: Duration) -> bool {
+    match last {
+        None => true,
+        Some(last_time) => now.duration_since(last_time).expect("time ran backwards") > cooldown,
+    }
+}
+
+fn is_enough_message_budget_left(context: &CycleContext<impl NetworkInterface>) -> bool {
+    context
+        .filtered_game_controller_state
+        .is_some_and(|game_controller_state| {
+            game_controller_state.remaining_number_of_messages
+                > context
+                    .spl_network
+                    .remaining_amount_of_messages_to_stop_sending
+        })
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Event {
+    None,
+    Striker(StrikerEvent),
```

Comment:
what about making this a structures enum variant?

Reply:
The inner type is also used as a function argument.